interfaces/builtin/pipewire_server.go

```diff
@@ -0,0 +1,151 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2023 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/interfaces/seccomp"
+	"github.com/snapcore/snapd/interfaces/udev"
+	"github.com/snapcore/snapd/snap"
+)
+
+// The pipewire-server interface is the companion interface to the audio-record
+// interface. The design of this interface is based on the idea that the slot
+// implementation (eg pulseaudio) is expected to query snapd on if the
+// audio-record slot is connected or not and the audio service will mediate
+// recording (ie, the rules below allow connecting to the audio service, but do
+// not implement enforcement rules; it is up to the audio service to provide
+// enforcement). If other audio recording servers require different security
+// policy for record and playback (eg, a different socket path), then those
+// accesses will be added to this interface.
+
+const pipewireServerSummary = `allows full access to the pipewire socket (don't needed for normal apps)`
+
+const pipewireServerBaseDeclarationSlots = `
+  pipewire-server:
+    allow-installation:
+      slot-snap-type:
+        - app
+        - core
+    deny-connection:
+      on-classic: false
+    deny-auto-connection: true
+`
+
+const pipewireServerConnectedPlugAppArmor = `
+# Allow communicating with pulseaudio service
+
+owner /{,var/}run/user/[0-9]*/ r,
+owner /{,var/}run/user/[0-9]*/pipewire-0 rwk,
+owner /{,var/}run/user/[0-9]*/pipewire-0.lock rwk,
+owner /{,var/}run/user/[0-9]*/pulse/pid rwk,
+`
+
+const pipewireServerConnectedPlugSecComp = `
+shmctl
+`
+
+const pipewireServerPermanentSlotAppArmor = `
+# When running Pipewire in system mode it will switch to the at
+# build time configured user/group on startup.
+capability setuid,
+capability setgid,
+
+capability sys_nice,
+capability sys_resource,
+
+# For udev
+network netlink raw,
+/sys/devices/virtual/dmi/id/sys_vendor r,
+/sys/devices/virtual/dmi/id/bios_vendor r,
+/sys/**/sound/** r,
+
+
+# Shared memory based communication with clients
```

Comment:
This comment doesn't seem to apply to any of the AppArmor statements. In the `audio_playback` interface this comment prefaces the statement:

`/{run,dev}/shm/pulse-shm-* mrwk,
`

One other related note... PulseAudio when run in system mode automatically disables shared memory support. Does PipeWire actually implement shared memory support for the Pulse native protocol, and if so, does it also disable when PipeWire (or more specifically the PipeWire-pulse daemon) run in system mode?

Reply:
But the idea is to run the code not in system mode, but as an user daemon, using

    apps:
      pipewire:
        command: snap/command-chain/pipewire
        daemon: simple
        passthrough:
          daemon-scope: user

Reply:
You can see the snap here: https://github.com/sergio-costas/pipewire-snap